### PR TITLE
Avalonia.Base - Styles improvements

### PR DESCRIPTION
**Pull request by Maruhl**
## What does the pull request do?
Primary OnCollectionChanged refactored (removed unnecessary cast and null checks in loops) 

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=8328) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=8328) before we can accept your contribution.<br/><br/>:x: Maruhl [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=8328)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=8328) it.</sub>

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021171-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021179-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Styling/Styles.cs [hunk 1]
```diff
@@ -227,70 +230,75 @@ void IResourceProvider.RemoveOwner(IResourceHost owner)
...
             }
+        }
 
-            void Remove(IList items)
+        private static void InternalRemove(IList items, IResourceHost? owner, ref StyleCache? cache)
+        {
+            if (owner is not null)
             {
-                for (var i = 0; i < items.Count; ++i)
+                foreach (var resourceProvider in items.OfType<IResourceProvider>())
```
Comment (hez2010) [inline src/Avalonia.Base/Styling/Styles.cs:267]: We generally avoid using LINQ to minimize allocations.
```suggestion
                for (var i = 0; i < items.Count; ++i)
                {
                    if (items[i] is IResourceProvider resourceProvider)
                    {
                        resourceProvider.RemoveOwner(owner);
                    }
                }
```

File: src/Avalonia.Base/Styling/Styles.cs [hunk 2]
```diff
@@ -227,70 +230,75 @@ void IResourceProvider.RemoveOwner(IResourceHost owner)
...
                 {
-                    var style = (IStyle)items[i]!;
+                    resourceProvider.RemoveOwner(owner);
```
Comment (hez2010) [inline src/Avalonia.Base/Styling/Styles.cs:269]: We generally avoid using LINQ to minimize allocations.

```suggestion
                    if (items[i] is IResourceProvider resourceProvider)
                    {
                        resourceProvider.RemoveOwner(owner);
                    }
```

Comment (Maruhl) [reply src/Avalonia.Base/Styling/Styles.cs]: Hello, thank you for taking the time to do the review.

Yes, you are basically right, if the method is called a million times then you should notice the difference.

That is again a code readability vs performance story.

If I search in the code of Avalonia for ".OfType<" I find 303 places :grin:.

Unfortunately I didn't find exact information what is standard here and what is not. At least I can't find anything about it at https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021181-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Styling/Styles.cs (hunk 2 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Styling/Styles.cs]: You wouldn't find Linq in perf critical places of the framework. Or if there are any - it should be fixed.

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0021208-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**